src/main/drivers/dma.h

```diff
@@ -108,8 +108,9 @@ typedef enum {
 
 #define DEFINE_DMA_IRQ_HANDLER(d, s, i) void DMA ## d ## _Stream ## s ## _IRQHandler(void) {\
                                                                 const uint8_t index = DMA_IDENTIFIER_TO_INDEX(i); \
-                                                                if (dmaDescriptors[index].irqHandlerCallback)\
-                                                                    dmaDescriptors[index].irqHandlerCallback(&dmaDescriptors[index]);\
+                                                                dmaCallbackHandlerFuncPtr handler = dmaDescriptors[index].irqHandlerCallback; \
```

Comment:
Using dummy/empty function may save 2 cycles (and few bytes of precious RAM) in common case

Reply:
So the non-inlined vtable call is actually faster than the null check?

Reply:
There was discussed here: https://github.com/betaflight/betaflight/issues/8535
The difference is hard to estimate on Cortex M4 and higher, branch predictor is involved.
When pointer is non-null, skipping test will save 1-3 cycles (branch is 1 cycle if not taken / 1+P if taken. Avoiding test may improve prediction on function call (1+P), saving up to 2 cycles)
In NULL case 1+P cycles is saved on test and 2x(1+P) added ... 

http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.ddi0439b/CHDDIGAC.html

Also using `-O3` may force gcc to duplicate a large part of code around `if` ...

Reply:
Also - no inlining and almost no optimization is possible when calling function through pointer - memory barrier is assumed (C standard allows called function to alter anything that is globally visible or accessible through arguments, even `const` one).
So all non-local memory cached in registers must be invalidated and reread after call.

Reply:
That still seems to be less than what is required to put the empty function onto the call stack and jump into it - but I guess it depends on how often we expect to be in the situation of referencing an empty function.

Reply:
> no inlining and almost no optimization is possible when calling function through pointer

That is kind of my point - we should avoid using vtables / function pointers in critical code where possible, and if this is not possible then null checking seems to be a better approach compared to incurring the cost of the non-optimisable call to an empty function through a vtable / function pointer.

Reply:
On ARM processor, empty function is only single instruction, `bx lr`. It takes 2-5 cycles. There is no call stack (ARM uses link register to store return address, called function stores it only when another function is called).

Null checking will be slower (1-3 cycles overhead) when function is called and faster when function is empty (2-6 cycles saved)

To get more optimization with `if`code, code inlining/duplication must take place, increasing code size, in some cases quite considerably.

In DMA case it makes little sense to use NULL as handler - code should simply disable IRQ mask for given channel. So using dummy function will save 1-3 cycles in almost all cases.

Reply:
In some cases, it may be considerably faster to use `switch` with list of functions that may be saved. ARM has good support for jump tables, maybe GCC will use them efficiently.
But it will remove a lot of abstraction from code ...

Reply:
On the F3, doing this saves 1 cycle:

`Dshot directionChange cycles: 377, micros: 5`.

I've added a commit, we can keep it or drop it as required.  One point to consider is veneers.  If the code for the DMA handler callback isn't close (in memory space) to the dma handler itself a veneer is used which is more cpu cycles.  Since the code also moves the dshot dma handler into the same memory segment as the dma handlers this isn't an issue.

See 97517e3644249436f57f66aca13fdca5f2ea0563

Reply:
The call is through register, so full address space is reachable.
And with tbh/tbb, it's on gcc (maybe lto will handle it, but I'm a bit sceptical)

Reply:
@ledvinap:

> In DMA case it makes little sense to use NULL as handler - code should simply disable IRQ mask for given channel. So using dummy function will save 1-3 cycles in almost all cases.

What is the reason to add a dummy function that uses up space but is never called?
As I said in #8535, in cases where calling the function pointer without it being set to an actual function is an error condition, adding a dummy function is just making this fact harder to understand as it gives the impression that the function pointer being called under these circumstances is not an error condition.

Reply:
there might be some merit in making debug builds cause a hard fault in the NoOp handler...

Reply:
@mikeller shall I remove the commit from this PR so that we can get the other changes merged?  We can continue discussion of this last issue here and I can make a new PR if you like?

personally I'm not overly fussed either way, the gains are minimal, however I like the zero-logic approach - i.e.  "just do it", even if "do it" is "nothing".  rather than "let me check if I should do this (every time), ok, let's do it".  In the case of the DMA IRQ handlers they should only be enabled if there is something to do, so the most-hit-case should be "do it".  if there's nothing to do there's either a hardware error (unlikely) or the hardware is misconfigured (likely during development).

Reply:
@hydra: I agree that null checking in a case where No-Op is indicative of an error condition is not the best approach (like I'd already stated in #8535), since we don't care about performance of the error case. But in this case the default handler should cause a hard fault, to avoid silently swallowing conditions that we assume are never met, and giving the impression that this is not an error case.

Reply:
ok, I'll add that to the PR in a bit then.  are you *sure* you don't want that hardfault in DEBUG builds only?

Reply:
@mikeller : (note that empty function wastes 4 bytes of FLASH - not a big deal.)

Maybe create global 'error' /'unexpected' /'fault'/'fail' function (prototype doesn't matter) that will either loop (`while(true)`), call hardfault handler or hardfault directly, possibly depending on DEBUG

Reply:
@ledvinap yeah, that sounds like a good idea. I'll see what I come up with.

Reply:
@ledvinap: Yeah, the space savings are minor. It's more the point that adding No-Op functions in case where the No-Op being called is actually indicative of an error makes the code harder to understand and debug - if we set the function pointer to something like `ErrorConditionHandler()` then it will become more apparent that this is a different pattern to the one where not calling any function is actually part of the normal control flow.

Reply:
the issue I'm currently having with this, is 
a) that the hardfault handler is defined in 'fc/hardfaults.c' and the dma.c is 'drivers/` code.  `drivers` code shouldn't call `fc` code.  suggestions on this welcomed.
b) causing a hardware seems wrong, and likely MCU dependent.
c) the is already duplication in the error handler code, this would add more.
d) the resulting error handler code LED handling should likely be refactored.

I'm going to drop the last commit from this PR so we can get this merged as it only saves 1 cycle and I don't _currently_ have to to invest further in this.

I pushed a branch with 3 commits, feel free to comment on the DMA handler changes there:

https://github.com/hydra/betaflight/commits/bf-dshot-telemetry-optimizations-2

I can open a PR for those changes if you guys like them, or simply for the original `dmaNoOpHandler` commit which a) works, b) doesn't have any of the issues in the above list.